crates/rover-client/src/blocking/client.rs

```diff
@@ -49,31 +52,61 @@ impl Client {
-            if !errs.is_empty() && errs[0].message.contains("406") {
-                return Err(RoverClientError::MalformedKey);
-            }
+        match response.status() {
+            StatusCode::OK => {
+                let response_body: graphql_client::Response<Q::ResponseData> = response.json()?;
 
-            return Err(RoverClientError::GraphQl {
-                msg: errs
-                    .into_iter()
-                    .map(|err| err.message)
-                    .collect::<Vec<String>>()
-                    .join("\n"),
-            });
-        }
+                if let Some(errs) = response_body.errors {
+                    if !errs.is_empty() && errs[0].message.contains("406") {
+                        return Err(RoverClientError::MalformedKey);
+                    }
+
+                    return Err(RoverClientError::GraphQl {
+                        msg: errs
+                            .into_iter()
+                            .map(|err| err.message)
+                            .collect::<Vec<String>>()
+                            .join("\n"),
+                    });
+                }
 
-        if let Some(data) = response_body.data {
-            Ok(data)
-        } else {
-            Err(RoverClientError::MalformedResponse {
-                null_field: "data".to_string(),
-            })
+                if let Some(data) = response_body.data {
+                    Ok(data)
+                } else {
+                    Err(RoverClientError::MalformedResponse {
+                        null_field: "data".to_string(),
+                    })
+                }
+            }
+            // This block specifically handles an error that is returned when
+            // Introspection is set to false on a production ApolloServer.
+            //
+            // We first check for a 400 HTTP Status Code (Bad Request). We then
+            // get the message sent by the server and display that to our users.
+            StatusCode::BAD_REQUEST => {
+                // It's not a given that an HTTP response is parsabel json,
+                // so let's match for a successful parse. Return a standard 400
+                // RoverClientError if we are unable to parse.
+                match response.json::<graphql_client::Response<Q::ResponseData>>() {
+                    Ok(body) => {
+                        if let Some(errs) = body.errors {
+                            return Err(RoverClientError::ClientError {
+                                msg: errs[0].message.to_string(),
+                            });
+                        }
+                        Err(RoverClientError::ClientError {
+                            msg: StatusCode::BAD_REQUEST.to_string(),
+                        })
+                    }
+                    Err(_) => Err(RoverClientError::ClientError {
+                        msg: StatusCode::BAD_REQUEST.to_string(),
+                    }),
```

Comment:
why would this be a bad request? seems like it would be closer to a malformed json response which we do have in `RoverClientError::InvalidJson` (not that I love that error, but it seems to make more sense here than returning a bad request error)

Reply:
I am very specifically matching on a 400 status code here, which is a `Bad Request`. If I am unable to get a response body (because not all 400 errors return response bodies), or if the body does not contain an error object, I would just rather return the status code that we received from this endpoint, which is `400 Bad Request`. I think anything else would be confusing to debug.